examples/pytorch/rgcn/utils.py

```diff
@@ -221,3 +220,104 @@ def calc_mrr(embedding, w, test_triplets, hits=[], eval_bz=100):
             avg_count = torch.mean((ranks <= hit).float())
             print("Hits (raw) @ {}: {:.6f}".format(hit, avg_count.item()))
     return mrr.item()
+
+#######################################################################
+#
+# Utility functions for evaluations (filtered)
+#
+#######################################################################
+
+def filter_o(triplets_to_filter, target_s, target_r, target_o, num_entities):
+    target_s, target_r, target_o = int(target_s), int(target_r), int(target_o)
+    filtered_o = []
+    # Do not filter out the test triplet, since we want to predict on it
+    if (target_s, target_r, target_o) in triplets_to_filter:
+        triplets_to_filter.remove((target_s, target_r, target_o))
+    # Do not consider an object if it is part of a triplet to filter
+    for o in range(num_entities):
+        if (target_s, target_r, o) not in triplets_to_filter:
+            filtered_o.append(o)
+    return torch.LongTensor(filtered_o)
+
+def filter_s(triplets_to_filter, target_s, target_r, target_o, num_entities):
+    target_s, target_r, target_o = int(target_s), int(target_r), int(target_o)
+    filtered_s = []
+    # Do not filter out the test triplet, since we want to predict on it
+    if (target_s, target_r, target_o) in triplets_to_filter:
+        triplets_to_filter.remove((target_s, target_r, target_o))
+    # Do not consider a subject if it is part of a triplet to filter
+    for s in range(num_entities):
+        if (s, target_r, target_o) not in triplets_to_filter:
+            filtered_s.append(s)
+    return torch.LongTensor(filtered_s)
+
+def perturb_o_and_get_filtered_rank(embedding, w, s, r, o, test_size, triplets_to_filter):
+    """ Perturb object in the triplets
+    """
+    num_entities = embedding.shape[0]
+    ranks = []
+    for idx in range(test_size):
+        if idx % 100 == 0:
+            print("test triplet {} / {}".format(idx, test_size))
+        target_s = s[idx]
+        target_r = r[idx]
+        target_o = o[idx]
+        filtered_o = filter_o(triplets_to_filter, target_s, target_r, target_o, num_entities)
+        target_o_idx = int((filtered_o == target_o).nonzero())
+        emb_s = embedding[target_s]
+        emb_r = w[target_r]
+        emb_o = embedding[filtered_o]
+        emb_triplet = emb_s * emb_r * emb_o
+        scores = torch.sigmoid(torch.sum(emb_triplet, dim=1))
+        _, indices = torch.sort(scores, descending=True)
+        rank = int((indices == target_o_idx).nonzero())
+        ranks.append(rank)
+    return torch.LongTensor(ranks)
+
+def perturb_s_and_get_filtered_rank(embedding, w, s, r, o, test_size, triplets_to_filter):
+    """ Perturb subject in the triplets
+    """
+    num_entities = embedding.shape[0]
+    ranks = []
+    for idx in range(test_size):
+        if idx % 100 == 0:
+            print("test triplet {} / {}".format(idx, test_size))
```

Comment:
You should move this for loop into calc_filtered_mrr. Otherwise there will be two print output of "test triplet {} / {}" in both perturb_s_and_get_filtered_rank and perturb_o_and_get_filtered_rank which is wired.

Reply:
Thank you for your suggestion. I would disagree that this is weird, since one print statement tracks the progress of perturbing the objects, and the other tracks the progress of perturbing subjects. This is made clear with print('Perturbing subject...') and print('Perturbing object...') in calc_filtered_mrr. The implementation of the raw MRR follows the same structure.

Reply:
Yep, adding "Perturbing subject .." and "Perturbing object .." is also another choice.